clang/test/CodeGenSYCL/device_global_int_footer_header.cpp

```diff
@@ -0,0 +1,112 @@
+// CHECK-HEADER-NEXT: };
+// CHECK-HEADER-NEXT: __sycl_device_global_registration __sycl_device_global_registerer;
+// CHECK-HEADER-NEXT: } // namespace
+// CHECK-HEADER: } // namespace detail
+// CHECK-HEADER: } // namespace sycl
+// CHECK-HEADER: } // __SYCL_INLINE_NAMESPACE(cl)
+
+// CHECK-FOOTER: #include <CL/sycl/detail/defines_elementary.hpp>
+
+// Shims go before the registration.
+// CHECK-FOOTER: namespace Foo {
+// CHECK-FOOTER-NEXT: namespace {
+// CHECK-FOOTER-NEXT: namespace __sycl_detail {
+// CHECK-FOOTER-NEXT: static constexpr decltype(AnonNS) &__shim_[[SHIM0:[0-9]+]]() {
+// CHECK-FOOTER-NEXT:   return AnonNS;
+// CHECK-FOOTER-NEXT: }
+// CHECK-FOOTER-NEXT: } // namespace __sycl_detail
+// CHECK-FOOTER-NEXT: } // namespace
+// CHECK-FOOTER-NEXT: } // namespace Foo
+// CHECK-FOOTER-NEXT: namespace {
+// CHECK-FOOTER-NEXT: namespace __sycl_detail {
+// CHECK-FOOTER-NEXT: static constexpr decltype(HasVarTemplate::VarTempl<int>) &__shim_[[SHIM1:[0-9]+]]() {
+// CHECK-FOOTER-NEXT:   return HasVarTemplate::VarTempl<int>;
+// CHECK-FOOTER-NEXT: }
+// CHECK-FOOTER-NEXT: } // namespace __sycl_detail
+// CHECK-FOOTER-NEXT: } // namespace
+
+// CHECK-FOOTER: #include <CL/sycl/detail/device_global_map.hpp>
+// CHECK-FOOTER: namespace sycl::detail {
+// CHECK-FOOTER-NEXT: namespace {
+// CHECK-FOOTER-NEXT: __sycl_device_global_registration::__sycl_device_global_registration() noexcept {
+
+device_global<int> Basic;
+// CHECK-FOOTER-NEXT: device_global_map::add((void *)&::Basic, "_Z5Basic");
+
+struct Wrapper {
+  static device_global<int> WrapperDevGlobal;
+};
+// CHECK-FOOTER-NEXT: device_global_map::add((void *)&::Wrapper::WrapperDevGlobal, "_ZN7Wrapper16WrapperDevGlobalE");
+
+template <typename T>
+struct WrapperTemplate {
+  static device_global<T> WrapperSpecID;
+};
+template class WrapperTemplate<int>;
+// CHECK-FOOTER-NEXT: device_global_map::add((void *)&::WrapperTemplate<int>::WrapperSpecID, "_ZN15WrapperTemplateIiE13WrapperSpecIDE");
+
+namespace Foo {
+device_global<int> NS;
+// CHECK-FOOTER-NEXT: device_global_map::add((void *)&::Foo::NS, "_ZN3Foo2NSE");
+
+inline namespace Bar {
+device_global<float> InlineNS;
+// CHECK-FOOTER-NEXT: device_global_map::add((void *)&::Foo::InlineNS, "_ZN3Foo3Bar8InlineNSE");
+
+struct Wrapper {
+  static device_global<int> WrapperDevGlobal;
+};
+// CHECK-FOOTER-NEXT: device_global_map::add((void *)&::Foo::Wrapper::WrapperDevGlobal, "_ZN3Foo3Bar7Wrapper16WrapperDevGlobalE");
+
+template <typename T>
+struct WrapperTemplate {
+  static device_global<T> WrapperDevGlobal;
+};
+template class WrapperTemplate<float>;
+// CHECK-FOOTER-NEXT: device_global_map::add((void *)&::Foo::WrapperTemplate<float>::WrapperDevGlobal, "_ZN3Foo3Bar15WrapperTemplateIfE16WrapperDevGlobalE");
+} // namespace Bar
+
+namespace {
+device_global<int> AnonNS;
+}
+// CHECK-FOOTER-NEXT: device_global_map::add((void *)&::Foo::__sycl_detail::__shim_[[SHIM0]](), "____ZN3Foo12_GLOBAL__N_16AnonNSE");
```

Comment:
Note to @Fznamznon : This command line in the check doesn't actually have a guid in it, so that might be why it doesn't show up, its probably worth giving that a shot.

Command line flag is -fsycl-unique-prefix=XXXX

The set of 4 underscores at the left side is I think associated with the unique prefix, the format is "<unique-prefix>____<rest>" (where those are 4 underscores).

Reply:
FWIW, I would think adding it to this test is probably beneficial.

Reply:
Ok, done.